### PR TITLE
FIX: Next Page issues with crawlers

### PR DESCRIPTION
**Pull request by jdmartinez1062**
**Context**
Currently, we have a monkey patch for Post voting plugin posts that allows us to load posts based a on different order than what `Post#sort_order` does. 

**Problem**
This is causing issues with crawlers (JS disabled) since we are displaying a `next_page` button that should not be there. This happens because the `next_page` logic is based on the supposition that the Post order is what `Post#sort_order` outputs. 

**Solution**
Make a `next_page` method that overrides core's implementation for Post voting Posts and works correctly with the custom order used in the plugin

### CODE REVIEW

File: spec/lib/topic_view_spec.rb [hunk 1]
```diff
@@ -213,6 +213,11 @@ def topic_view_near(post)
         [answer_minus_2_votes.id, answer_minus_1_vote.id, answer_0_votes.id],
       )
     end
+    describe "#next_page" do
+      it "should return the next page properly when the highest id post is not the last" do
+        expect(TopicView.new(topic.id, user, { post_number: post.post_number }).next_page).to eql(2)
```
Comment (CvX) [inline spec/lib/topic_view_spec.rb:218]:
```suggestion

    describe "#next_page" do
      it "returns the next page properly when the highest id post is not the last" do
        expect(TopicView.new(topic.id, user, { post_number: post.post_number }).next_page).to eq(2)
```

File: extensions/topic_view_extension.rb [hunk 2]
```diff
@@ -19,6 +19,26 @@ def self.prepended(base)
...
+      return super if !topic.is_post_voting?
+      return super if @filter == TopicView::ACTIVITY_FILTER
+
+      @highest_post_number = @filtered_posts.last.post_number
+
+      if @posts.blank?
+        @last_post = nil
+      else
+        @last_post ||= @posts.last
+      end
```
Comment (CvX) [inline extensions/topic_view_extension.rb:32]: Do you need this or can you reuse `last_post`, like the core implementation? https://github.com/discourse/discourse/blob/140b2118af26bbf013bde31fa37ce1c0a62c77b7/lib/topic_view.rb#L188-L202

File: extensions/topic_view_extension.rb [hunk 3]
```diff
@@ -19,6 +19,26 @@ def self.prepended(base)
...
+
+      @next_page ||=
+        begin
+          if @last_post && @highest_post_number && (@highest_post_number != @last_post.post_number)
+            @page + 1
+          end
+        end
```
Comment (CvX) [inline extensions/topic_view_extension.rb:39]: 🏌️ 

```suggestion
        if @last_post && @highest_post_number && (@highest_post_number != @last_post.post_number)
          @page + 1
        end
```

File: extensions/topic_view_extension.rb [hunk 4]
```diff
@@ -19,6 +19,26 @@ def self.prepended(base)
       end
     end
 
+    def next_page
+      return super if !topic.is_post_voting?
+      return super if @filter == TopicView::ACTIVITY_FILTER
```
Comment (tgxworld) [inline extensions/topic_view_extension.rb:24]: I think we can extract `!topic.is_post_voting? || @filter == TopicView::ACTIVITY_FILTER` to another method so we don't duplicate it across two methods.

**Review by markvanlan — APPROVED** [review]

File: extensions/topic_view_extension.rb [hunk 5]
```diff
@@ -82,5 +81,21 @@ def filter_posts_near(post_number)
...
+      @highest_post_number = @filtered_posts.last.post_number
+
+      @next_page ||=
+        if last_post && @highest_post_number && (@highest_post_number != last_post.post_number)
+          @page + 1
+        end
+    end
+
+    def post_voting_topic?
+      return false if !topic.is_post_voting? || @filter == TopicView::ACTIVITY_FILTER
```
Comment (markvanlan) [inline extensions/topic_view_extension.rb:97]: just 
```
def post_voting_topic?
  topic.is_post_voting? && @filter != TopicView::ACTIVITY_FILTER
end
```

**Pull request merged**